libsolidity/interface/CompilerStack.cpp

```diff
@@ -922,8 +921,29 @@ string CompilerStack::createMetadata(Contract const& _contract) const
 			meta["sources"][s.first]["urls"].append("bzzr://" + toHex(s.second.swarmHash().asBytes()));
 		}
 	}
-	meta["settings"]["optimizer"]["enabled"] = m_optimize;
-	meta["settings"]["optimizer"]["runs"] = m_optimizeRuns;
+	/// Backwards compatibility
+	/// TODO The point of this is backwards compatibility, but people might misunderstand
+	/// this to be a "master override".
+	meta["settings"]["optimizer"]["enabled"] =
+	(
+		m_optimiserSettings.runDeduplicate &&
+		m_optimiserSettings.runCSE &&
+		m_optimiserSettings.runOrderLiterals &&
+		m_optimiserSettings.runConstantOptimiser
+	);
+	solAssert(m_optimiserSettings.expectedExecutionsPerDeployment < std::numeric_limits<Json::LargestUInt>::max(), "");
+	meta["settings"]["optimizer"]["runs"] = Json::Value(Json::LargestUInt(m_optimiserSettings.expectedExecutionsPerDeployment));
+
+	meta["settings"]["optimizer"]["orderLiterals"] = m_optimiserSettings.runOrderLiterals;
+	meta["settings"]["optimizer"]["jumpdestRemover"] = m_optimiserSettings.runJumpdestRemover;
+	meta["settings"]["optimizer"]["peephole"] = m_optimiserSettings.runPeephole;
+	meta["settings"]["optimizer"]["deduplicate"] = m_optimiserSettings.runDeduplicate;
+	meta["settings"]["optimizer"]["cse"] = m_optimiserSettings.runCSE;
+	// We used "z" in the past in the standard-json, so we use "constantOptimizer".
+	meta["settings"]["optimizer"]["constantOptimizer"] = m_optimiserSettings.runConstantOptimizer;
+	meta["settings"]["optimizer"]["yul"] = m_optimiserSettings.runYulOptimizer;
```

Comment:
The idea is to add another field that specifies detailed instructions for the yul optimiser. In the settings, we could check that we do not have the combination of "yulSettings: something" and "yul: false"